### DISCUSSION TITLE
Operation returned an invalid status code 'Unauthorized' while trying to do json patch  for custom resource objects

### DISCUSSION THREAD

**Discussion by srinivasmummareddy** [kubernetes-client/csharp#1026] [Q&A]
The following is the stack trace of exception

"Error","@x":"k8s.Autorest.HttpOperationException: Operation returned an invalid status code 'Unauthorized'\n   at k8s.Kubernetes.SendRequestRaw(String requestContent, HttpRequestMessage httpRequest, CancellationToken cancellationToken)\n   at k8s.AbstractKubernetes.k8s.ICustomObjectsOperations.PatchNamespacedCustomObjectWithHttpMessagesAsync(Object body, String group, String version, String namespaceParameter, String plural, String name, String dryRun, String fieldManager, Nullable`1 force, IReadOnlyDictionary`2 customHeaders, CancellationToken cancellationToken)\n   at k8s.CustomObjectsOperationsExtensions.PatchNamespacedCustomObjectAsync(ICustomObjectsOperations operations, Object body, String group, String version, String namespaceParameter, String plural, String name, String dryRun, String fieldManager, Nullable`1 force, CancellationToken cancellationToken)

Note: The particular pod has a service account which has permission to patch the crd. 

Not able to figure out the exact reason

I'm using  Kubernetes client : 8.0.68

**Comment by tg123** [comment]
are you using `BuildDefaultConfig`?

**Reply by srinivasmummareddy** [reply]
No, We are using config file

` KubernetesClientConfiguration.BuildConfigFromConfigFile(kubeconfigPath: kubeConfigFileName);`

**Reply by tg123** [reply]
inside or outside the pod?
looks like your config is not correct.
what about kubectl

**Reply by srinivasmummareddy** [reply]
@tg123  Its inside the pod. I am able to do all the other kubernetes operations like creating secrets and configmaps from this pod. Only I'm facing issue with patch of crd. (Service account is having permissions for patching as well)


My bad: the following is the client code

   KubernetesClientConfiguration config;
            if (KubernetesClientConfiguration.IsInCluster())
            {
                config = KubernetesClientConfiguration.InClusterConfig();
            }
            else
            {
                // use k3s kubeconfig if it exists, if not, fall back to default config path
                string? kubeConfigFileName = File.Exists(K3sKubeConfigFileName) ? K3sKubeConfigFileName : null;
                config = KubernetesClientConfiguration.BuildConfigFromConfigFile(kubeconfigPath: kubeConfigFileName);
            }

            _client = new k8s.Kubernetes(config);

**Reply by tg123** [reply]
please use `BuildDefaultConfig` inside pod
it will fill config automatically

**Reply by srinivasmummareddy** [reply]
@tg123  It didn't work still I'm getting the same issue.
Note I'm able to do all other operations

The following is the error


"k8s.Autorest.HttpOperationException: Operation returned an invalid status code 'Unauthorized'\n   at k8s.Kubernetes.SendRequestRaw(String requestContent, HttpRequestMessage httpRequest, CancellationToken cancellationToken)\n   at k8s.AbstractKubernetes.k8s.ICustomObjectsOperations.PatchNamespacedCustomObjectWithHttpMessagesAsync(Object body, String group, String version, String namespaceParameter, String plural, String name, String dryRun, String fieldManager, Nullable`1 force, IReadOnlyDictionary`2 customHeaders, CancellationToken cancellationToken)\n   at k8s.CustomObjectsOperationsExtensions.PatchNamespacedCustomObjectAsync(ICustomObjectsOperations operations, Object body, String group, String version, String namespaceParameter, String plural, String name, String dryRun, String fieldManager, Nullable`1 force, CancellationToken cancellationToken)\n

**Reply by tg123** [reply]
hmm i do not think you have

could you please double confirm with `can i`?
see doc here
https://kubernetes.io/docs/reference/access-authn-authz/authorization/#checking-api-access

**Reply by srinivasmummareddy** [reply]
@tg123 yes, I cross-checked that a few days back. Also tried installing kubectl client manually inside pod and patched the crd :) its able to do that. Not exactly sure whats the issue

**Reply by tg123** [reply]
could you please check like below?
```
kubectl auth can-i list pods \
	--namespace target \
	--as system:serviceaccount:dev:dev-sa
```

**Reply by srinivasmummareddy** [reply]
![image](https://user-images.githubusercontent.com/22257824/192228136-d47414fe-578e-43c8-b7f5-e2e019d23de6.png)

I did this test earlier with the patch operation as well.. You can see it is having access to listing pods. 
Let me know if you need more details

**Reply by tg123** [reply]
i mean patch your crd

**Reply by srinivasmummareddy** [reply]
![image](https://user-images.githubusercontent.com/22257824/192229941-6f2c901d-7442-4cbb-b4fc-4c40823d6432.png)

**Reply by tg123** [reply]
hmm next step would be checking logs of api server or capture http pkg

**Reply by srinivasmummareddy** [reply]
Can you elaborate more or give some commands to do this

**Reply by tg123** [reply]
https://kubernetes.io/docs/tasks/debug/debug-cluster/#control-plane-nodes
check api server to verify what user is actually using

you can also use audit
https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/

**Comment by tg123** [comment]
check if it helps
https://github.com/kubernetes-client/csharp/issues/971#issuecomment-1260189357

**Reply by srinivasmummareddy** [reply]
@tg123  nope its giving the same error...no luck

**Reply by srinivasmummareddy** [reply]
https://github.com/kubernetes-client/csharp/issues/624
Is it something interlinked? Even I'm getting a similar issue not the exact scenario but pattern is bit similar
I'm getting unauthorized using c# client but its working fine with kubectl

**Reply by tg123** [reply]
did you capture net pkg? any detailed response from server side?

**Reply by srinivasmummareddy** [reply]
not seeing any audit logs while doing that operation using c# sharp client. If I'm doing the same with kubectl I'm able to see the audit logs and patch operation is working
Note: I have executed that particular patch command  using same service account used in my pod
ex: kubectl --as system:serviceaccount:security:security-sa patch keycloakrealms kdiedge --type "json" -p="[{\"op\": \"replace\", \"path\":\"/spec/realm/users/0/credentials/0/value\", \"value\":\"srinivas\"}]" -n keycloak
